analyses/cell-type-ewings/exploratory_analysis/05-cluster-exploration.Rmd

```diff
@@ -0,0 +1,672 @@
+      dplyr::filter(library_id == id) |>
+      ggplot(aes(x = UMAP1, y = UMAP2, color = cluster)) +
+      geom_point(alpha = 0.5, size = 0.1) +
+      facet_wrap(vars(nn_param)) +
+      theme(
+        aspect.ratio = 1
+      ) +
+      labs(title = id) +
+      guides(color = guide_legend(override.aes = list(alpha = 1, size = 1.5)))
+  })
+```
+
+### Clustering statistics 
+
+Below we calculate a series of statistics: 
+
+- Average silhouette width: This metric evaluates cluster separation. 
+Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
+- Average cluster purity: This metric also evaluates cluster separation and tells us the proportion of neighboring cells that are assigned to the same cluster. 
+Higher purity values indicate that clusters are well separated. 
+- Cluster stability: This evaluates how stable the clustering is to input data. 
+Higher values of cluster stability indicate more reproducible clusters. 
+
+```{r, warning=FALSE}
+# get a combined stats dataframe with purity and width for all clusters
+all_stats_df <- sce_list |>
+  purrr::imap(\(sce, library_id){
+    get_cluster_stats(sce, library_id, all_cluster_results)
+  }) |>
+  dplyr::bind_rows(.id = "library_id") |>
+  dplyr::mutate(
+    # make sure the order is correct
+    nn_param = forcats::fct_relevel(nn_param, "k_5", after = 0)
+  )
+```
+
+
+```{r}
+# silhouette width for different params
+library_ids |>
+  purrr::map(\(id){
+    plot_cluster_stats(all_stats_df, width, id)
+  })
+```
+
+```{r}
+# cluster purity for different params
+library_ids |>
+  purrr::map(\(id){
+    plot_cluster_stats(all_stats_df, purity, id)
+  })
+```
+
+Here we see that both silhouette width and cluster purity peak around k=20 or k=25. 
+For `SCPCL000822`, increasing k after 25 does not change the width or purity, 
+For `SCPCL000824`, increasing k after 20-25 decreases the width and keeps purity around the same. 
+
+```{r, warning=FALSE, message=FALSE}
+# calculate cluster stability
+stability_df <- sce_list |>
+  purrr::imap(\(sce, id){
+    get_cluster_stability(sce, id, all_cluster_results)
+  }) |>
+  dplyr::bind_rows(.id = "library_id") |>
+  dplyr::mutate(
+    k_value = as.factor(k_value),
+    # make sure that k = 5 comes first
+    k_value = forcats::fct_relevel(k_value, "5", after = 0)
+  )
+
+# plot stability across all values of k
+ggplot(stability_df, aes(x = k_value, y = ari)) +
+  geom_jitter(width = 0.1) +
+  facet_grid(rows = vars(library_id)) +
+  labs(title = "Cluster stability") +
+  stat_summary(
+    aes(group = k_value),
+    color = "red",
+    # median and quartiles for point range
+    fun = "median",
+    fun.min = function(x) {
+      quantile(x, 0.25)
+    },
+    fun.max = function(x) {
+      quantile(x, 0.75)
+    }
+  )
+```
+
+Here we see that cluster stability increases as k increases for both libraries. 
+We see a plateau start to appear around k = 20 and k = 25. 
```

Comment:
Looking at this again, I actually now wonder if we should be at k=30/35 for `SCPCL000824` with this metric? What do you think?